[Merged by Bors] - feat(topology/homotopy/homotopy_group): `group` and `comm_group` instances for `π_n`

ralvrz opened the pull request:
This PR adds:
- Group instance for `π_(n+1)`
- Commutative group instance for `π_(n+2)`

---

port: https://github.com/leanprover-community/mathlib4/pull/3874

- [x] depends on: #16879

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ralvrz committed:
feat(topology/homotopy): define nth homotopy group πₙ, show equivalences of π₀ and π₁

ralvrz committed:
fix line too long

ralvrz committed:
add missing docstrings

ralvrz committed:
fix curly braces lint

ralvrz committed:
fix broken proof

ralvrz committed:
add missing instances

ralvrz committed:
fix

ralvrz committed:
add missing docstring

ralvrz committed:
fix style

ralvrz committed:
rewrite proofs

ralvrz committed:
renamings and congruence proofs

ralvrz committed:
add missing type annotations

ralvrz committed:
renamings

ralvrz committed:
remove indentations

alreadydone committed:
golf + docstring

alreadydone committed:
slight golf

ralvrz committed:
Merge branch 'pi_n_golfs' into pi_n

ralvrz committed:
add missing docstring

ralvrz committed:
rename file

ralvrz committed:
Apply suggestions from code review

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

ralvrz committed:
unique instance of I^0

ralvrz committed:
remove one_indep, change fin 1 → α with I^1

ralvrz committed:
Apply suggestions from code review

Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

alreadydone committed:
feat(logic/unique): functions from a `unique` type is `const` (#14823)

+ A function `f` from a `unique` type is equal to the constant function with value `f default`, and the analogous heq version for dependent functions.

+ Also changes `Π a : α, Sort v` in the file to `α → Sort v`.

Inspired by https://github.com/leanprover-community/mathlib/pull/14724#discussion_r900542203

ralvrz committed:
one_char by eq_const_of_unique

ralvrz committed:
group instance sketch

ralvrz committed:
group instance proofs

ralvrz committed:
progress in path equivalence

ralvrz committed:
progress in homotopy equivalence

ralvrz committed:
cleaning

ralvrz committed:
merge and derived instances

ralvrz committed:
reorder

ralvrz committed:
merge fix

ralvrz committed:
missing assumptions

ralvrz committed:
homotopic transported backwards

ralvrz committed:
feat(topology/compact_open):  continuous_comp left functor C(-, γ) (#15068)



Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ralvrz committed:
fix lines too long

ralvrz committed:
missing docstrings

ralvrz committed:
missing assumptions

ralvrz committed:
inline stuff

ralvrz committed:
docstrings

ralvrz committed:
minor fixes

ralvrz committed:
generalized index by induction

ralvrz committed:
comm_group instance

ralvrz committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into pi_n_testing

ralvrz committed:
simp and linter fixes

ralvrz committed:
golfs

ralvrz committed:
fix long line

ralvrz committed:
instances

ralvrz committed:
Merge branch 'pi_n' of github.com:leanprover-community/mathlib into pi_n

ralvrz committed:
Merge branch 'pi_n_testing' into pi_n

alexjbest commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -193,3 +535,137 @@ begin
+
+end
+
+/-- Concatecantion forms a group.-/
```
```suggestion
/-- Concatenation forms a group.-/
```

alexjbest commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -193,3 +535,137 @@ begin
+lemma is_unital : @eckmann_hilton.is_unital (π_(n+2) x) (+) 1 :=
+⟨⟨const_concat 1⟩,⟨concat_const 1⟩⟩
+
+/-- Conmutativity of horizontal concatenation is shown by
```
```suggestion
/-- Commutativity of horizontal concatenation is shown by
```

ralvrz committed:
Apply suggestions from code review

Co-authored-by: Alex J Best <alex.j.best@gmail.com>

alreadydone committed:
WIP

alreadydone committed:
golfed is_comm_group

alreadydone committed:
concat_eq

alreadydone committed:
bunch of comments/TODO + golf!

alreadydone committed:
more golfs

alreadydone committed:
supply proof: pi.locally_compact_space

alreadydone committed:
comment on pi instances

alreadydone committed:
add_comm_group

alreadydone committed:
locally_compact_space.pi, finte case

ralvrz committed:
cleaning

kbuzzard committed:
remove two non-terminal dsimps

kbuzzard committed:
manually squeezed another dsimp

kbuzzard committed:
squeezed last dsimp!

alreadydone committed:
remove squeeze dsimp TODO, add missing argument

ralvrz committed:
feat(topology/subset_properties): `locally_compact_space` instance for `Π` types (#15707)

This PR adds
- `locally_compact_space.pi` mirroring `locally_compact_space.prod` and
- `locally_compact_space.pi_finite` for finite products

Proof by: @alreadydone

Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

ralvrz committed:
docstrings and linter fixes

ralvrz committed:
equivalente to logic.equiv.basic

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -270,6 +258,7 @@ begin
+/--Homotopy of paths to `C(I × cube N, X)`. -/
 @[simps] def homotopy_from (i : N) {p q : gen_loop N x}
   (H : (to_path i p).homotopy (to_path i q)) : C(I × cube N, X) :=
 ((⟨_,continuous_map.continuous_uncurry⟩ : C(_,_)).comp
```
```suggestion
((⟨_, continuous_map.continuous_uncurry⟩ : C(_,_)).comp
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -386,9 +378,13 @@ end
 
+instance : group (π_(n+1) X x) := is_group
+
+/--Additive group on `π_(n+2)`. -/
```
```suggestion
/-- Another group structure on `π_(n+2)` that distributes over the default one, so as to enable the Eckmann-Hilton argument. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -35,111 +44,284 @@ noncomputable theory
+section
+variable [decidable_eq N]
+
+lemma insert_boundary (i : N) {t₀ : I} {t} (H : (t₀ = 0 ∨ t₀ = 1) ∨ t ∈ boundary {j // j ≠ i}) :
```
needs to be renamed

alreadydone commented on src/logic/equiv/basic.lean:
```diff
@@ -817,6 +817,15 @@ product over `α` of `β (some α)` and `β none` -/
+/-- The dependant product of a type family `β : α → Type` is equivalent to the pair of `β a` and
+the dependant product of all but `a`, for any `a:α` -/
+@[simps] def pi_equiv_at_prod_not_at {α : Type*} [decidable_eq α] {β : α → Type*} (a : α) :
+   (Π x, β x) ≃ (β a) × (Π a', a' ≠ a → β a') :=
```
I'm not sure why you switched from subtype to `Π a', a' ≠ a`? This seems harder to use.

alreadydone replied:
The following works:
```
@[simps] def pi_equiv_single_prod_erase {α : Type*} [decidable_eq α] (i : α) (β : α → Type*) :
  (Π j, β j) ≃ β i × Π j : {j // j ≠ i}, β j :=
{ to_fun := λ f, ⟨f i, λ j, f j⟩,
  inv_fun := λ f j, if h : j = i then h.symm.rec f.1 else f.2 ⟨j, h⟩,
  right_inv := λ f, by { ext, exacts [dif_pos rfl, (dif_neg x.2).trans (by cases x; refl)] },
  left_inv := λ f, by { ext, dsimp only, split_ifs, { subst h }, { refl } } }
```
You can make a shorter definition using existing constructions:
```
(equiv.pi_equiv_pi_subtype_prod _ β).trans $
  (equiv.Pi_subsingleton (λ j : {j // j = i}, β j) ⟨i, rfl⟩).prod_congr $ equiv.refl _ 
```
and `@[simps]` would generate the same simp lemmas (as confirmed by `@[simps?]`, but it slows down `@[simps]` from 600-700ms to ~3s in the homeomorph version, so I don't go for it.

I've opened PR #16210 with the first definition, and specializes it to get a non-dependent version. The `eq.rec` would be annoying, but the `@[simps]` automatically removes it in the generated `simp` lemma, so I think it won't cause any trouble.

I don't include `merge_split_self` in the PR because it's only used once and can be easily inlined.

alreadydone replied:
#16210 has been merged. I'm polishing the PR and will push shortly.

alreadydone reviewed (commented):
Hi @ralvrz have you been preparing any further preliminary PR or done any work cleaning up this PR recently? I'm about to PR `merge_split_equiv` (dependent version, both equiv and homeomorph) in a few days if you're not planning to do so.

alreadydone committed:
Merge remote-tracking branch 'origin/master' into pi_n

alreadydone committed:
WIP

alreadydone committed:
implicit

alreadydone committed:
clean up a bit

alreadydone committed:
Merge remote-tracking branch 'origin/master' into pi_n

ralvrz committed:
merge fix

ralvrz committed:
split_at abbreviation and symm_spec

ralvrz committed:
rename and fixes

ralvrz committed:
update docstrings

ralvrz committed:
removes homeomorph edit

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -47,7 +46,8 @@ universes u
 
-/-- The `n`-dimensional cube. -/
+/-- The `n`-dimensional cube `I^n` is defined via an abstract definition so that subsets of faces
+  can be implemented with subtypes of `fin n`. -/
```
```suggestion
/-- We allow an arbitrary indexing type `N`, not just `fin n`, in the definition of `cube`, so as to
  be able to identify facets of a cube to cubes indexed by subtypes of `N`, which is more
  convenient than using `fin (n-1)`. -/
```

alreadydone commented on src/topology/homeomorph.lean:
```diff
@@ -504,7 +504,6 @@ variables [decidable_eq ι] (i : ι)
   continuous_inv_fun := continuous_pi $ λ j, by { dsimp only [equiv.pi_split_at],
     split_ifs, subst h, exacts [continuous_fst, (continuous_apply _).comp continuous_snd] } }
 
-variable (β)
```
Why are you removing this? It allows us to write `fun_split_at I i` rather than `@fun_split_at I _ _ _ i`, and should be easy to port to mathlib4.

ralvrz replied:
I did it to remove the tag in case that made it easier to close this PR, I'll just revert it since it was already easy to port.
Should the abbreviations stay? I think the abbreviation for the inverse makes some lemmas easier to parse for a human (eg. `symm_spec`)

alreadydone replied:
Yes, I am fine with keeping the `split_at`, `insert_at` abbreviations.

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -378,12 +396,10 @@ begin
   simp_rw [if_neg fin.zero_ne_one, if_neg fin.zero_ne_one.symm],
-  split_ifs; { congr, ext1, apply ite_ite_comm, sorry },
+  split_ifs; { congr, ext1, apply ite_ite_comm,
+  intros h0 h1, rw h0 at h1, exact zero_ne_one (fin.veq_of_eq h1) },
```
```suggestion
  split_ifs; { congr, ext1, apply ite_ite_comm, rintro rfl, exact fin.zero_ne_one },
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -74,8 +74,15 @@ lemma one_char (f : I^1) : f = λ _, f 0 := eq_const_of_unique f
-lemma insert_boundary (i : N) {t₀ : I} {t} (H : (t₀ = 0 ∨ t₀ = 1) ∨ t ∈ boundary {j // j ≠ i}) :
-  (fun_split_at I i).symm ⟨t₀, t⟩ ∈ boundary N :=
+/-- Splits a cube into a pair I × {j // j≠i} → I such that the first member is the value of the
+  cube at coordinate `i`, and the function returns the value of the cube elsewhere. -/
```
I think this is clear enough:
```suggestion
/-- The forward direction of the homeomorphism
  between the cube $I^N$ and $I × I^{N\setminus\{j\}}$. -/
```
Your docstring has two main problems:
+ I would say "split a cube into a **product**", "split **(the coordinates) of a point** in a cube into a pair".
+ `I × {j // j≠i} → I` needs a pair of parentheses: `I × ({j // j≠i} → I)`.

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -74,8 +74,15 @@ lemma one_char (f : I^1) : f = λ _, f 0 := eq_const_of_unique f
+  cube at coordinate `i`, and the function returns the value of the cube elsewhere. -/
+abbreviation split_at (i : N) := @fun_split_at I _ _ _ i
+
+/-- Makes a cube from a pair I × {j // j≠i} → I, inverse of `split_at`. -/
```
```suggestion
/-- The backward direction of the homeomorphism
  between the cube $I^N$ and $I × I^{N\setminus\{j\}}$. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -92,7 +99,8 @@ local notation `Ω` := loop_space
 
-/-- The `n`-dimensional generalized loops; functions `I^n → X` fixed at the boundary. -/
+/-- The `n`-dimensional generalized loops are continuous functions `I^n → X` fixed at the boundary,
+  defined with an abstract type `N` of coordinates. -/
```
I think this is clearer:
```suggestion
/-- The `n`-dimensional generalized loops based at `x` in a space `X` are
  continuous functions `I^n → X` that sends the boundary to `x`.
  We allow an arbitrary indexing type `N` in place of `fin n` here. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -140,18 +148,18 @@ section
-    λ y yH, p.property ((fun_split_at I i).symm (t, y)) (cube.insert_boundary i $ or.inr yH)⟩,
+{ to_fun := λ t, ⟨(p.val.comp (cube.insert_at i).to_continuous_map).curry t,
+    λ y yH, p.property (cube.insert_at i (t, y)) (cube.insert_at_boundary i $ or.inr yH)⟩,
   continuous_to_fun := by continuity,
```
automatically filled
```suggestion
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -140,18 +148,18 @@ section
-/-- Generalized loop from a path by `extrac`-ing of `I×I^n`. -/
+/-- Generalized loop from a path by uncurrying `I × {j // j ≠ i} → I` into `cube N`. -/
 @[simps] def from_path (i : N) : Ω (gen_loop {j // j ≠ i} x) const → gen_loop N x :=
 λ p, ⟨(⟨λ t, (p t).1, by continuity⟩ : C(I, C(cube _, X))).uncurry.comp
```
```suggestion
@[simps] def from_path (i : N) (p : Ω (gen_loop {j // j ≠ i} x) const) : gen_loop N x :=
⟨(⟨λ t, (p t).1, by continuity⟩ : C(I, C(cube _, X))).uncurry.comp
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -140,18 +148,18 @@ section
 variable [decidable_eq N]
 
-/-- Path from a generalized loop by `insert`-ing into `I^(n+1)`. -/
+/-- Path from a generalized loop by currying `cube N` into `I × {j // j ≠ i} → I`. -/
```
This is easier to understand to me:
```suggestion
/-- Path from a generalized loop by currying $I^N → X$ into $I → (I^{N\setminus\{j\}} → X)$. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -140,18 +148,18 @@ section
+  target' := by { ext t, refine p.property (cube.insert_at i (1, t)) ⟨i, or.inr _⟩, simp } }
 
-/-- Generalized loop from a path by `extrac`-ing of `I×I^n`. -/
+/-- Generalized loop from a path by uncurrying `I × {j // j ≠ i} → I` into `cube N`. -/
```
```suggestion
/-- Generalized loop from a path by uncurrying $I → (I^{N\setminus\{j\}} → X)$ into $I^N → X$. -/
```

alreadydone reviewed (changes requested):
Hi @ralvrz, thanks for coming back to this and thanks for the cleanup and timeout fixes! (Edit: it looks like some timeouts just disappeared, e.g. the `@[simps]` timeout in `homotopy_to`.) I think this PR is pretty much ready to go but I have some issues with some docstrings, and some micro-optimizations.

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,292 @@ Authors: Roberto Alvarez
-that `π 1 x` is equivalent to the fundamental group at `x`.
+We show that `π_0 x` is equivalent to the path-conected components, and
+that `π_1 x` is equivalent to the fundamental group at `x`.
+We give a group instance using path transitivity and show commutativity when `n > 1`.
```
This is more familiar math terminology, I think.
```suggestion
We give a group instance using path composition and show commutativity when `n > 1`.
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,292 @@ Authors: Roberto Alvarez
+  insert_at i ⟨t₀, t⟩ ∈ boundary N :=
+begin
+  cases H, { use i, rwa [fun_split_at_symm_apply, dif_pos rfl] },
+  cases H with j H,
```
```suggestion
  obtain H | ⟨j, H⟩ := H,
  { use i, rwa [fun_split_at_symm_apply, dif_pos rfl] },
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,292 @@ Authors: Roberto Alvarez
+def const : gen_loop N x := ⟨continuous_map.const _ x, λ _ _, rfl⟩
 
-instance inhabited : inhabited (gen_loop n x) := { default := const }
+@[simp] lemma const_eq {t} : (@const X _ N x) t = x := rfl
```
```suggestion
@[simp] lemma const_apply {t} : (@const X _ N x) t = x := rfl
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,292 @@ Authors: Roberto Alvarez
+    function.uncurry_apply_pair],
+  by_cases Heq : j = i,
+  { subst Heq, cases Hj; rw Hj; simp only [p.source, p.target]; convert const_eq },
+  { exact gen_loop.boundary _ _ ⟨⟨j, Heq⟩, Hj⟩ },
```
```suggestion
  obtain rfl | Hne := eq_or_ne j i,
  { cases Hj; rw Hj; simp only [p.source, p.target]; convert const_apply },
  { exact gen_loop.boundary _ _ ⟨⟨j, Hne⟩, Hj⟩ },
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,292 @@ Authors: Roberto Alvarez
+end
+
+/-- The `n+1`-dimensional loops are equivalent to the loop space at `const` as a `n`-dimensional
+  loop, defined abtractically for any nonempty index type `N`. -/
```
```suggestion
/-- The `n+1`-dimensional loops are in bijection with the loops in the space of
  `n`-dimensional loops with base point `const`.
  We allow an arbitrary indexing type `N` in place of `fin n` here. -/
```
(We may also put the last sentence in the module docstring so we don't need to repeat it in every declaration. BTW, the adverb of "abstract" is "abstractly".)

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,292 @@ Authors: Roberto Alvarez
+
+section
+
+/--Coercion as a continuous map.-/
```
```suggestion
/-- The inclusion from the space of generalized loops to the space of all continuous functions
  (not necessarily constant on the boundary), as a continuous map. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,292 @@ Authors: Roberto Alvarez
+
+variable [decidable_eq N]
+
+/-- Composition with insert as a continuous map.-/
```
```suggestion
/-- Composition with insert as a continuous map. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,292 @@ Authors: Roberto Alvarez
+⟨λ f, f.comp (cube.insert_at i).to_continuous_map,
+  (cube.insert_at i).to_continuous_map.continuous_comp_left⟩
+
+/--Homotopy of generalized loops to `C(I × I, C(cube {j // j ≠ i}, X))`. -/
```
```suggestion
/-- A homotopy between `n+1`-dimensional loops `p` and `q` constant on the boundary
  seen as a continuous map from a square to the space of `n`-dimensional paths. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,292 @@ Authors: Roberto Alvarez
+@[simps] def path_equiv (i : N) : gen_loop N x ≃ Ω (gen_loop {j // j ≠ i} x) const :=
+{ to_fun := to_path i,
+  inv_fun := from_path i,
+  left_inv := λ p, by { ext, exact congr_arg p (equiv.apply_symm_apply _ _) },
```
```suggestion
  left_inv := λ p, by { ext, exact congr_arg p (equiv.right_inv _ _) },
```
and three more occurrences in the proof of `homotopic_from`.

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -193,3 +343,63 @@ begin
+lemma from_path_trans_to_path {p q : gen_loop N x} (i : N) {t} :
+  (path_equiv i).symm ((path_equiv i p).trans $ path_equiv i q) t = if (t i : ℝ) ≤ 1/2
+    then p (λ j, if j = i then set.proj_Icc 0 1 zero_le_one (2 * t i) else t j)
+    else q (λ j, if j = i then set.proj_Icc 0 1 zero_le_one (2 * t i - 1) else t j) :=
```
I wrote this originally but I now think we should probably make the RHS a `def` of type `gen_loop N x` (maybe called `mul_at i p q`) with continuity transferred from the LHS using this lemma. Then we can state `mul_spec` (for all `i : N`) using the new `def` rather than using an existential.

ralvrz replied:
I've also written a `def` for `symm_spec` in the same vein, I could use some help golfing the proofs.
To state `mul_spec` for all `i : N` we need the group instance using `i` instead of `0 :  fin (n+1)` but using `i` the instance can't be synthesized, so I left it in terms of `0 : fin (n+1)`.

alreadydone replied:
I pushed the generalizations to all `i : N` and changed your definitions to use `gen_loop.copy` which I introduced. If you are happy with it, I'll send it to maintainers soon!

ralvrz replied:
Looks fine to me 👍

alreadydone replied:
Thanks! However as I took a final look I found some more problems in docstrings and went on to improve them. Currently, there is a final [naming issue](https://github.com/leanprover-community/mathlib/pull/15681/files#r1189178897) that remains. Do you think we should change the names?

alreadydone reviewed (commented):
Some more comments; I think these are all I have!

ralvrz committed:
Apply suggestions from code review

Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

ralvrz committed:
more suggestions

ralvrz committed:
explicit argument to ¹fun_split_at¹

ralvrz committed:
specs to defs

alreadydone committed:
generalize mul/inv_spec, mul_at→trans_at & lemmas

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
-`homotopic` relation.
--/
+/-- The `n`th homotopy group at `x` defined as the quotient of `gen_loop n x` by the
+  `homotopic` relation. -/
```
```suggestion
  `gen_loop.homotopic` relation. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+    apply congr_arg p <|> apply congr_arg q,
+    exact (equiv.right_inv _ _).symm },
+end
+-- above proofs: still room for golfing?
```
```suggestion
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+  refine nonempty.map (λ H, ⟨⟨⟨λ t, ⟨homotopy_to i H t, _⟩, _⟩, _, _⟩, _⟩),
+  { rintros y ⟨i,iH⟩,
+    rw homotopy_to_apply_apply, rw H.eq_fst, rw p.2,
+    all_goals { apply cube.insert_at_boundary, right, exact ⟨i,iH⟩} },
```
```suggestion
  { rintros y ⟨i, iH⟩,
    rw homotopy_to_apply_apply, rw H.eq_fst, rw p.2,
    all_goals { apply cube.insert_at_boundary, right, exact ⟨i, iH⟩} },
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+
+variable [decidable_eq N]
+
+/-- Composition with insert as a continuous map. -/
```
```suggestion
/-- Composition with `cube.insert_at` as a continuous map. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
-f.to_continuous_map.homotopic_rel g.to_continuous_map (cube.boundary n)
+instance inhabited : inhabited (gen_loop N x) := ⟨const⟩
+
+/-- The "homotopy relative to boundary" relation between `gen_loop`s. -/
```
```suggestion
/-- The "homotopic relative to boundary" relation between `gen_loop`s. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+  (cube.insert_at i).to_continuous_map.continuous_comp_left⟩
+
+/-- A homotopy between `n+1`-dimensional loops `p` and `q` constant on the boundary
+  seen as a continuous map from a square to the space of `n`-dimensional paths. -/
```
```suggestion
  seen as a homotopy between two paths in the space of `n`-dimensional paths. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
--/
-structure gen_loop (n : ℕ) (x : X) extends C(I^n, X) :=
-(boundary : ∀ y ∈ cube.boundary n, to_fun y = x)
+/-- Paths fixed at both ends -/
```
```suggestion
/-- The space of paths with both endpoints equal to a specified point `x : X`. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+  exacts [H.apply_zero _, H.apply_one _],
+end
+
+/--Homotopy of paths to `C(I × cube N, X)`. -/
```
```suggestion
/-- The converse to `gen_loop.homotopy_to`: a homotopy between two loops in the space of `n`-dimensional loops can be seen as a homotopy between two `n+1`-dimensional paths. -/
```

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+variable [decidable_eq N]
+
+/-- Path from a generalized loop by currying $I^N → X$ into $I → (I^{N\setminus\{j\}} → X)$. -/
+@[simps] def to_path (i : N) : gen_loop N x → Ω (gen_loop {j // j ≠ i} x) const := λ p,
```
This should probably be renamed `to_loop` since the codomain is the loop space, and same for `from_path`. Then we should systematically change "path" to "loop" in the docstrings, except when there is no restriction on boundary behavior.

ralvrz replied:
Done!

alreadydone replied:
Thanks! 

maintainer merge

alreadydone committed:
Apply suggestions from code review

ralvrz committed:
rename path to loop

eric-wieser commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+  Examples with 𝕊^n (π_n (𝕊^n) = ℤ, π_m (𝕊^n) trivial for m < n).
+  Actions of π_1 on π_n.
+  Group (up to homotopy) of Ω.
+  Lie algebra: ⁅π_(n+1), π_(m+1)⁆ contained in π_(n+m+1).
```
```suggestion
TODO:
* Path-induced homomorphisms. Show that `pi1_equiv_fundamental_group` is a group isomorphism.
* Examples with `𝕊^n (π_n (𝕊^n) = ℤ`, `π_m (𝕊^n)` trivial for `m < n`.
* Actions of π_1 on π_n.
* Group (up to homotopy) of Ω.
* Lie algebra: `⁅π_(n+1), π_(m+1)⁆` contained in `π_(n+m+1)`.
```

eric-wieser commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+/-- The points in a cube with at least one projection equal to 0 or 1. -/
+def boundary (N) : set (cube N) := {y | ∃ i, y i = 0 ∨ y i = 1}
 
-/--
-The projection to the last `n` coordinates from an `n+1` dimensional cube.
--/
-@[simp] def tail {n} : I^(n+1) → I^n := λ c, fin.tail c
```
Is there a reason you removed this?

ralvrz replied:
It wasn't being used anywhere, and it's just a synonym for `fin.tail`

eric-wieser commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+
+/-- The forward direction of the homeomorphism
+  between the cube $I^N$ and $I × I^{N\setminus\{j\}}$. -/
+abbreviation split_at (i : N) := fun_split_at I i
```
Please add the type of this declaration and the one below
```suggestion
abbreviation split_at (i : N) : sorry := fun_split_at I i
```

eric-wieser commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+
+/-- The backward direction of the homeomorphism
+  between the cube $I^N$ and $I × I^{N\setminus\{j\}}$. -/
+abbreviation insert_at (i : N) := (fun_split_at I i).symm
```
I would recommend using `@[reducible] def` instead of `abbreviation`, as the former does not generate equation lemmas, leading to worse docs.

(throughout)

eric-wieser commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+abbreviation loop_space (X : Type*) [topological_space X] (x : X) := path x x
+local notation `Ω` := loop_space
+
+instance loop_space.inhabited : inhabited (Ω X x) := ⟨path.refl x⟩
```
Does `by apply_instance` work here? If so, you can just delete this.

ralvrz replied:
`apply_instance` fails, but the instance isn't used anyway so I'll delete it.

alreadydone replied:
I think it's nice to have this instance, and I'm surprised that the nonempty_instance linter doesn't complain about the removal. I guess it's because it's a reducible def and because [path](https://leanprover-community.github.io/mathlib_docs/topology/path_connected.html#path) has `nolint has_nonempty_instance` attribute.

eric-wieser replied:
In that case we should be adding the instance for `path` directly.

alreadydone replied:
Maybe not worth modifying one more ported file. Maybe @ralvrz will add a TODO item after port?

alreadydone replied:
Added
```
instance loop_space.inhabited : inhabited (path x x) := ⟨path.refl x⟩
```
in the same file.

eric-wieser commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
+
+/-- The inclusion from the space of generalized loops to the space of all continuous functions
+  (not necessarily constant on the boundary), as a continuous map. -/
+abbreviation c_coe : C(gen_loop N x, C(cube N, X)) := ⟨λ p, p.val, continuous_induced_dom⟩
```
Can `library_search` find an existing definition to use here? This should hold in general for any subtype, I think.

If not, are you sure you want this to be an abbreviation?

ralvrz replied:
`library_search` fails, it's only used once but in-lining the definition gives errors.

alreadydone replied:
> This should hold in general for any subtype, I think.

I think so as this is just a bundled (and specialized) version of [continuous_induced_dom](https://leanprover-community.github.io/mathlib_docs/topology/order.html#continuous_induced_dom), but we probably don't want to make a bundled version for every declaration in that file ... It may be possible to inline it with proper type annotation and/or using `continuous_map.comp` in `homotopy_from`.

YaelDillies commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
-def cube (n : ℕ) : Type := fin n → I
-local notation `I^` := cube
+def cube (N : Type*) := N → I
+local notation `I^` n := cube (fin n)
```
Sorry, I know you haven't introduced that, but why have a type synonym for pi types?

ralvrz replied:
Essentially, just for legibility, like `N×M` matrices over `X` being functions `N → M → X`.

YaelDillies replied:
Yes, but there are other ways of improving legibility than introducing new definitions. For example, reducible definitions or notation.

alreadydone replied:
Actually, @ralvrz introduced `cube` in #14724. It's originally defined to be `fin n → I` as you can see, but now that it's simply `N → I` I think we can just remove it, or make it reducible (so no need of `@[derive]`), or extend the notation `I^` to allow arbitrary `N` (it's not used many times in this file currently).

ralvrz replied:
I've removed the def and generalized the `I^` notation for arbitrary `N`.
Similarly, using `Ω^` for `gen_loop` is suggestive even if not actually an exponential..

alreadydone replied:
Yeah, making `Ω^N` a local notation seems reasonable; indeed it's common to write `Ω^n` for iterated loop spaces as in https://en.wikipedia.org/wiki/Spectrum_(topology)#Suspension_spectrum etc., but since we don't want to write `Ω^fin n` it's wise to keep the `Ω^N` notation local, so we can introduce the `Ω^n` notation when needed.

ralvrz committed:
Apply suggestions from code review

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ralvrz committed:
More suggestions

eric-wieser commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,306 @@ Authors: Roberto Alvarez
--/
+/-- We allow an arbitrary indexing type `N`, not just `fin n`, in the definition of `cube`, so as to
+  be able to identify facets of a cube to cubes indexed by subtypes of `N`, which is more
+  convenient than using `fin (n-1)`. -/
```
This docstring change has discarded the *what* part of the docstring and left only the *why*.

alreadydone replied:
Maybe we'll remove this def ...

YaelDillies replied:
Yeah I think it's better to remove the def. I checked and was quite baffled to see that `_root_.cube` was a thing.

ralvrz replied:
Done!

eric-wieser commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,306 @@ Authors: Roberto Alvarez
+
+/-- The inclusion from the space of generalized loops to the space of all continuous functions
+  (not necessarily constant on the boundary), as a continuous map. -/
+@[reducible] def c_coe : C(gen_loop N x, C(cube N, X)) := ⟨λ p, p.val, continuous_induced_dom⟩
```
We're missing a more general definition here,
```lean
/-- `subtype.val` as a continuous function -/
def continuous_map.subtype {α} (p : α → Prop) [topological_space α] : C(subtype p, α) :=
{ to_fun := coe, continuous_to_fun := continuous_induced_dom }
```
which belongs in an earlier file.

alreadydone replied:
Probably not worth modifying another ported file, given that `c_coe` is [easily inlined](https://github.com/leanprover-community/mathlib/pull/15681/commits/c953156ebaed677e80cd14d3ad0c026349e259dd#diff-8d28e1c85b22df3c99f61de0a89bc1e7ba35a56eae6db74bcdec15750d007e50L250-R241). (The continuity tactic filling in the continuity field automatically makes it less useful to have the bundled version.)

eric-wieser commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,306 @@ Authors: Roberto Alvarez
--/
-def boundary (n : ℕ) : set (I^n) := {y | ∃ i, y i = 0 ∨ y i = 1}
+instance compact_space : compact_space (cube N) :=
+by { convert pi.compact_space, intro, apply_instance }
```
This indicates we're missing an workaround instance; we should have `function.compact_space` too, right next to `pi.compact_space`.

alreadydone replied:
Done! (modifies one more ported file)

eric-wieser commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -193,3 +355,103 @@ begin
+
+lemma is_unital_aux_group (i : N) :
+  eckmann_hilton.is_unital (aux_group i).mul (⟦const⟧ : homotopy_group N X x) :=
+⟨⟨(aux_group i).one_mul⟩, ⟨(aux_group i).mul_one⟩⟩
```
I think you should be able to prove this with [`eckmann_hilton.mul_one_class.is_unital`](https://leanprover-community.github.io/mathlib_docs/group_theory/eckmann_hilton.html#eckmann_hilton.mul_one_class.is_unital)

alreadydone replied:
It won't be shorter though; either `@eckmann_hilton.mul_one_class.is_unital _ (aux_group i).to_monoid.to_mul_one_class` or `by { letI := aux_group i, exact eckmann_hilton.mul_one_class.is_unital }`.

eric-wieser replied:
That second spelling looks good to me.

alreadydone replied:
Turns out a type ascription is needed; if you think it's worth it, I'll commit. I don't want to abuse `;` but otherwise it doesn't fit in one line.
```suggestion
by letI : group (homotopy_group N X x) := aux_group i; exact eckmann_hilton.mul_one_class.is_unital
```

eric-wieser commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -193,3 +355,103 @@ begin
+def trans_at (i : N) (f g : gen_loop N x) : gen_loop N x :=
+copy ((loop_equiv i).symm ((loop_equiv i f).trans $ loop_equiv i g))
+  (λ t, if (t i : ℝ) ≤ 1/2
+    then f (λ j, if j = i then set.proj_Icc 0 1 zero_le_one (2 * t i) else t j)
```
```suggestion
    then f (function.update t i (set.proj_Icc 0 1 zero_le_one (2 * t i)))
```
is a bit more idiomatic

ralvrz committed:
suggestion

ralvrz committed:
removes  def

ralvrz committed:
notation for gen_loop

alreadydone committed:
remove c_coe, function dot notation, instances

alreadydone committed:
missing space

fpvandoorn commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,301 @@ Authors: Roberto Alvarez
+
+/-- A homotopy between `n+1`-dimensional loops `p` and `q` constant on the boundary
+  seen as a homotopy between two paths in the space of `n`-dimensional paths. -/
+def homotopy_to (i : N) {p q : Ω^N x} (H : p.1.homotopy_rel q.1 (cube.boundary N)) :
```
```suggestion
def homotopy_to (i : N) {p q : Ω^N x} (H : p.homotopic q) :
```
?
Also a few times below

alreadydone replied:
We are transforming data here, so the input needs to be data. `homotopic_to` is the corresponding Prop, whose proof relies on specific properties of the transformed data (behavior at endpoints), not just existence.

fpvandoorn commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,301 @@ Authors: Roberto Alvarez
+def homotopy_group (N) (X : Type*) [topological_space X] (x : X) : Type _ :=
+quotient (gen_loop.homotopic.setoid N x)
+/-- Homotopy group of finite index. -/
+@[reducible] def pi (n) (X : Type*) [topological_space X] (x : X) := homotopy_group (fin n) _ x
```
I believe this is currently not in a namespace? It should be in one

fpvandoorn commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,301 @@ Authors: Roberto Alvarez
 
 /-- The 0-dimensional generalized loops based at `x` are in 1-1 correspondence with `X`. -/
-def gen_loop_zero_equiv : gen_loop 0 x ≃ X :=
+def gen_loop_zero_equiv : Ω^(fin 0) x ≃ X :=
```
Generalize `fin 0` to any empty type

fpvandoorn commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -154,42 +314,140 @@ begin
-    (p.boundary (λ _, 0) ⟨0, or.inl rfl⟩)
-    (p.boundary (λ _, 1) ⟨1, or.inr rfl⟩),
+/-- The 1-dimensional generalized loops based at `x` are in 1-1 correspondence with loops at `x`. -/
+@[simps] def gen_loop_one_equiv_loop : Ω^(fin 1) x ≃ Ω X x :=
```
This should be a consequence of `gen_loop.loop_equiv` and `gen_loop_zero_equiv` (generalized to empty types) and functoriality of `Ω`

fpvandoorn replied:
Also generalize this to types with the `unique` condition.

alreadydone replied:
> This should be a consequence of `gen_loop.loop_equiv` and `gen_loop_zero_equiv` (generalized to empty types) and functoriality of `Ω`

I refrained from doing this suggested refactor because we don't have the functoriality yet; I added it as a TODO item. Generalization to `unique` is done.

fpvandoorn commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -154,42 +314,140 @@ begin
+/-- The first homotopy group at `x` is equivalent to the fundamental group, i.e. the loops based at
+  `x` up to homotopy. -/
+-- TODO: deduce from homotopy_group_equiv_fundamental_group?
+def pi1_equiv_fundamental_group : π_ 1 X x ≃ fundamental_group X x :=
```
generalize to `unique` types

fpvandoorn replied:
After redefining `gen_loop_one_equiv_loop` the fact that it preserves `homotopy` should follow from earlier results.

fpvandoorn replied:
Oh even better: generalize `pi0_equiv_path_components` to empty types and use that + `homotopy_group_equiv_fundamental_group`

alreadydone replied:
Generalizations are done but I refrained from the refactoring because we should probably first combine [category_theory.functor.map_Aut](https://leanprover-community.github.io/mathlib_docs/category_theory/endomorphism.html#category_theory.functor.map_Aut) and [fundamental_groupoid.fundamental_groupoid_functor](https://leanprover-community.github.io/mathlib_docs/algebraic_topology/fundamental_groupoid/basic.html#fundamental_groupoid.fundamental_groupoid_functor) in another file, which is out of scope of this PR.

alreadydone replied:
@fpvandoorn I've done the refactoring in another branch anyway, and you can evaluate its (de)merits by [comparing both branches](https://github.com/leanprover-community/mathlib/compare/pi_n...homotopy_group_equiv_fundamental_group_of_unique_refactor?expand=1). Although defeq works nicely so that no `path.cast` is required, `@[simps]` times out at `gen_loop_equiv_of_unique` on the new branch. The whole refactoring saves 3 lines (not counting the lines added in path_connected that are not actually used).

fpvandoorn commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,301 @@ Authors: Roberto Alvarez
+quotient (gen_loop.homotopic.setoid N x)
+/-- Homotopy group of finite index. -/
+@[reducible] def pi (n) (X : Type*) [topological_space X] (x : X) := homotopy_group (fin n) _ x
+local notation `π_` := pi
```
This should be `localized` notation (same for all other `local notation`)

alreadydone replied:
Will be using locale `topology`.

fpvandoorn commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -154,42 +314,140 @@ begin
+variables {n : ℕ} (i : N)
+
+/-- Group structure on `π_(n+1)`. -/
+instance group : group (π_(n+1) X x) :=
```
Is there a reason this is not an instance on `homotopy_group`?

EDIT: Oh, `N` doesn't have a canonical element. Ignore this comment.

fpvandoorn replied:
This should be on `π_n` assuming `[ne_zero n]` 

It should probably also exist for any type `N` with element `n`, but that requires a new variant of `homotopy_group` that takes `n` as (unused) argument, so this should probably wait for a later PR.

alreadydone replied:
We could use `[inhabited N]` here. Or even `[nonempty N]` and use choice, since the end result doesn't depend on the choice by `mul_spec`.

fpvandoorn replied:
Yeah, that is probably useful. And indeed, the use of `nonempty` seems reasonable here.

alreadydone replied:
Generalized to arbitrary `nonempty` type.

fpvandoorn commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -154,42 +314,140 @@ begin
+end
+
+/-- Characterization of multiplicative identity -/
+lemma const_spec : (1 : π_(n+1) X x) = ⟦const⟧ := rfl
```
```suggestion
lemma const_def : (1 : π_(n+1) X x) = ⟦const⟧ := rfl
```
(feel free to ignore) I think we tend to use `_def` not `_spec` (also below)

eric-wieser replied:
`foo_spec` is normally reserved for `p (foo x)` where `p y` is hard to name (and certainly not when it's `y = ...`)

alreadydone replied:
Maybe `one_def`? `⟦const⟧` is not the LHS but the RHS, and it's not `const` itself but its equivalence class.

I do want to keep the names `mul_spec` and `inv_spec` because those are not just unfolding definitions but also involve showing the equivalence class is independent of `i : N`. Their purposes are to show the multiplication and inversion operations defined by pulling back the operations on the fundamental group is the same as the usual (e.g. [Wikipedia](https://wikimedia.org/api/rest_v1/media/math/render/svg/d292ba059108f5c62b4abe9edb1bb2776cb22cf8)) definition.

fpvandoorn commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -154,42 +314,140 @@ begin
+by { rw [symm_at_indep 0 p, ← from_loop_symm_to_loop], apply quotient.sound, refl }
+
+/-- Multiplication on `π_(n+2)` is commutative. -/
+instance comm_group : comm_group (π_(n+2) X x) :=
```
This would also be nice to have for `π_n` with a type-class assumption that `n > 1`. Not sure if we already have such a class, so this can wait for later.

alreadydone replied:
I don't think we have such a class but we could use [nontrivial](https://leanprover-community.github.io/mathlib_docs/logic/nontrivial.html#nontrivial) in a statement for arbitrary type (not just `fin`). Again, we could use either (`nonempty` + choice) or `inhabited`.

alreadydone replied:
Generalized to `nontrivial` types.

fpvandoorn commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,301 @@ Authors: Roberto Alvarez
 
-TODO: show that `π n x` is a group when `n > 0` and abelian when `n > 1`. Show that
-`pi1_equiv_fundamental_group` is an isomorphism of groups.
+TODO:
```
(some version of) `Ω^n (Ω^m X) \equiv Ω^(n+m) X` is also a good target

alreadydone replied:
Should probably use `sum N M`. And separately `Ω^N X \equiv Ω^M X` given `N \equiv M`.

alreadydone replied:
Added
```
* `Ω^M (Ω^N X) ≃ Ω^(M⊕N) X`, and `Ω^M X ≃ Ω^N X` when `M ≃ N`. Similarly for `π_`.
```
as a TODO item.

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,301 @@ Authors: Roberto Alvarez
-def pi0_equiv_path_components : π 0 x ≃ zeroth_homotopy X :=
+/-- The 0th homotopy "group" is equivalent to the path components of `X`, aka the `zeroth_homotopy`.
+  -/
+def pi0_equiv_path_components : π_ 0 X x ≃ zeroth_homotopy X :=
```
should say `zeroth_homotopy` rather than `path_components` by naming convention

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,301 @@ Authors: Roberto Alvarez
+-- Should be generated with `@[simps]` but it times out.
+lemma homotopy_to_apply (i : N) {p q : Ω^N x} (H : p.1.homotopy_rel q.1 $ cube.boundary N)
+  (t : I × I) (tₙ : I^{j // j ≠ i}) :
+    homotopy_to i H t tₙ = H (t.fst, cube.insert_at i (t.snd, tₙ)) := rfl
```
```suggestion
  homotopy_to i H t tₙ = H (t.fst, cube.insert_at i (t.snd, tₙ)) := rfl
```
mathlib3 style

alreadydone reviewed (commented):
Thanks for the review! I'll implemented suggested changes by the end of the weekend if @ralvrz doesn't beat me to it.

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,141 +5,301 @@ Authors: Roberto Alvarez
+* Path-induced homomorphisms. Show that `pi1_equiv_fundamental_group` is a group isomorphism.
+* Examples with `𝕊^n (π_n (𝕊^n) = ℤ`, `π_m (𝕊^n)` trivial for `m < n`.
+* Actions of π_1 on π_n.
+* Group (up to homotopy) of Ω.
```
This is now in mathlib as [path.H_space](https://leanprover-community.github.io/mathlib_docs/topology/homotopy/H_spaces.html#path.H_space) right? Just removed it.
```suggestion
* Group (up to homotopy) of Ω.
```

alreadydone committed:
address review

alreadydone committed:
const_spec -> one_def

alreadydone committed:
loop_space.inhabited

alreadydone committed:
inhabited -> nonempty

alreadydone committed:
cosmetics

alreadydone committed:
cosmetics

alreadydone committed:
Upgrade loop_equiv to loop_homeo (just for fun)

alreadydone committed:
cosmetics

alreadydone committed:
Make x explicit in gen_loop_homeo_of_is_empty

alreadydone committed:
outdated docstring

alreadydone committed:
Namespace homotopy_group -> gen_loop for some declarations

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,191 +5,462 @@ Authors: Roberto Alvarez
-TODO: show that `π n x` is a group when `n > 0` and abelian when `n > 1`. Show that
-`pi1_equiv_fundamental_group` is an isomorphism of groups.
+TODO:
+* Functoriality of `Ω`.
```
```suggestion
```

alreadydone committed:
This TODO is done with `path.map_map` etc.

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,191 +5,461 @@ Authors: Roberto Alvarez
-`pi1_equiv_fundamental_group` is an isomorphism of groups.
+TODO:
+* `Ω^M (Ω^N X) ≃ₜ Ω^(M⊕N) X`, and `Ω^M X ≃ₜ Ω^N X` when `M ≃ N`. Similarly for `π_`.
+* Path-induced homomorphisms. Show that `pi1_equiv_fundamental_group` is a group isomorphism.
```
```suggestion
* Path-induced homomorphisms. Show that `homotopy_group.pi_1_equiv_fundamental_group`
  is a group isomorphism.
```

alreadydone committed:
Renamed declaration in TODO.

alreadydone commented on src/topology/homotopy/homotopy_group.lean:
```diff
@@ -5,191 +5,462 @@ Authors: Roberto Alvarez
-  to the boundary
+* `gen_loop N x` is the type of continuous fuctions `I^N → X` that send the boundary to `x`,
+* `homotopy_group.pi n x` denoted `π_ n X x` is the quotient of `gen_loop (fin n) x` by homotopy
+  relative to the boundary,
```
```suggestion
* `homotopy_group.pi n X x` denoted `π_ n X x` is the quotient of `gen_loop (fin n) x` by
  homotopy relative to the boundary,
```

alreadydone committed:
docstring

alreadydone commented on src/topology/subset_properties.lean:
```diff
@@ -1131,6 +1133,12 @@ instance locally_compact_space.pi [∀ i, compact_space (π i)] : locally_compac
+  locally_compact_space (ι → β) := locally_compact_space.pi_finite
+
+instance locally_compact_space.function [locally_compact_space β] [compact_space β] :
+  locally_compact_space (ι → β) := locally_compact_space.pi
```
Should we call these `function.locally_compact_space_of_finite`, `function.locally_compact_space`? We should then also rename the three existing instances ([locally_compact_space.pi_finite](https://leanprover-community.github.io/mathlib_docs/topology/subset_properties.html#locally_compact_space.pi_finite) etc.) similarly.

YaelDillies replied:
Indeed, the existing instances are misnamed.

alreadydone replied:
Let me change them and see if mathlib compiles. (Yes it does!)

alreadydone committed:
Fix instance names

alreadydone committed:
Merge branch 'master' into pi_n
